// angular-bootstrap-toasts/esm5/lib/Components/toast-message/toast-message.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ToastMessage } from '../../Models/toast-message.models';
import { AngularBootstrapToastsService } from '../../angular-bootstrap-toasts.service';
import { filter, tap } from 'rxjs/operators';
import { interval } from 'rxjs';
var ToastMessageComponent = /** @class */ (function () {
    function ToastMessageComponent(toastsService) {
        this.toastsService = toastsService;
        this.progressLineWidth = '100%';
        this.msInterval = 10;
        this.isMouseFocused = false;
    }
    /**
     * @return {?}
     */
    ToastMessageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.currentDuration = this.Toast.Duration;
        this.durationTimer = interval(this.msInterval).pipe(filter((/**
         * @return {?}
         */
        function () { return !_this.isMouseFocused; })), tap((/**
         * @return {?}
         */
        function () {
            _this.currentDuration -= _this.msInterval;
            if (_this.Toast.IsProgressLineEnabled) {
                _this.progressLineWidth = (100 / _this.Toast.Duration * _this.currentDuration).toFixed(0) + '%';
            }
            if (_this.currentDuration <= 0) {
                _this.Toast.Close();
            }
        }))).subscribe();
    };
    /**
     * @return {?}
     */
    ToastMessageComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.durationTimer) {
            this.durationTimer.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    ToastMessageComponent.prototype.animationPause = /**
     * @return {?}
     */
    function () {
        this.isMouseFocused = true;
    };
    /**
     * @return {?}
     */
    ToastMessageComponent.prototype.animationPlay = /**
     * @return {?}
     */
    function () {
        this.isMouseFocused = false;
    };
    /**
     * @return {?}
     */
    ToastMessageComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        if (this.Toast.IsClosableByMouseClick) {
            this.Toast.Close(false);
        }
    };
    /**
     * @param {?=} confirmationResult
     * @return {?}
     */
    ToastMessageComponent.prototype.remove = /**
     * @param {?=} confirmationResult
     * @return {?}
     */
    function (confirmationResult) {
        this.Toast.Close(confirmationResult || false);
    };
    ToastMessageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Toast-Message',
                    template: "<div\r\n    class=\"_toast {{ Toast.ToastClass || '' }}\"\r\n    (mouseover)=\"animationPause()\"\r\n    (mouseout)=\"animationPlay()\"\r\n    (click)=\"onClick()\">\r\n    <div class=\"_toast-header {{ Toast.TitleClass || '' }}\">\r\n        <i\r\n            *ngIf=\"Toast.IconClass\"\r\n            class=\"{{ Toast.IconClass }}\"></i>\r\n\r\n        <strong\r\n            class=\"mr-auto\">{{ Toast.Title }}</strong>\r\n\r\n        <small *ngIf=\"Toast.Moment\">{{ Toast.Moment }}</small>\r\n\r\n        <button\r\n            type=\"button\"\r\n            class=\"_close {{ Toast.CloseButtonClass || '' }}\"\r\n            (click)=\"remove()\">\r\n            <span aria-hidden=\"true\">&times;</span>\r\n        </button>\r\n    </div>\r\n\r\n    <div\r\n        *ngIf=\"Toast.IsProgressLineEnabled\"\r\n        class=\"_duration-line {{ Toast.ProgressLineClass || '' }}\"\r\n        [style.width]=\"progressLineWidth\"></div>\r\n\r\n    <div\r\n        class=\"_toast-body {{ Toast.BodyClass || '' }}\"\r\n        [innerHTML]=\"Toast.Text\"></div>\r\n\r\n    <div\r\n        *ngIf=\"Toast.Type === 'confirm' && Toast.ToolbarItems\"\r\n        class=\"_toast-toolbar {{ Toast.ToolbarClass || '' }}\">\r\n        <button\r\n            *ngIf=\"Toast.ToolbarItems.actionButton.visible\"\r\n            class=\"_toast-toolbar_actionButton {{ Toast.ToolbarItems.actionButton.class || 'btn btn-success btn-sm' }}\"\r\n            (click)=\"remove(true)\">{{ Toast.ToolbarItems.actionButton.text || 'Confirm' }}</button>\r\n\r\n        <button\r\n            *ngIf=\"Toast.ToolbarItems.cancelButton.visible\"\r\n            class=\"_toast-toolbar_cancelButton {{ Toast.ToolbarItems.cancelButton.class || 'btn btn-light btn-sm' }}\"\r\n            (click)=\"remove(false)\">{{ Toast.ToolbarItems.cancelButton.text || 'Cancel' }}</button>\r\n    </div>\r\n</div>",
                    styles: [":host{display:block}._toast{width:100%;margin-bottom:10px;border:1px solid #dadada;background:#fff;box-shadow:0 7px 9px -4px rgba(0,0,0,.2),0 14px 21px 2px rgba(0,0,0,.14),0 5px 26px 4px rgba(0,0,0,.12)}._toast-header{width:100%;display:flex;align-items:center;padding:8px;border-bottom:1px solid #dfe0e1;background:#f5f5f5;color:#6c757d}._duration-line{height:3px;background:rgba(0,0,0,.25);transition:.1s}._toast-body{font-size:12px;padding:12px;background:#fff}._toast-header i{margin-right:7px}._toast-header strong{font-size:16px}._toast-header small{font-size:11px;color:#6c757d}._toast-toolbar{margin:10px}._toast-toolbar_actionButton{margin-right:8px}button._close{padding:0;margin-left:.5rem;margin-bottom:.25rem;float:right;-webkit-appearance:none;-moz-appearance:none;appearance:none;background-color:transparent;border:0;font-size:1.5rem;font-weight:700;line-height:1;color:#000;opacity:.6}button._close:hover{cursor:pointer;opacity:.8}"]
                }] }
    ];
    /** @nocollapse */
    ToastMessageComponent.ctorParameters = function () { return [
        { type: AngularBootstrapToastsService }
    ]; };
    ToastMessageComponent.propDecorators = {
        Toast: [{ type: Input }]
    };
    return ToastMessageComponent;
}());
export { ToastMessageComponent };
if (false) {
    /** @type {?} */
    ToastMessageComponent.prototype.Toast;
    /** @type {?} */
    ToastMessageComponent.prototype.currentDuration;
    /** @type {?} */
    ToastMessageComponent.prototype.progressLineWidth;
    /**
     * @type {?}
     * @private
     */
    ToastMessageComponent.prototype.msInterval;
    /**
     * @type {?}
     * @private
     */
    ToastMessageComponent.prototype.isMouseFocused;
    /**
     * @type {?}
     * @private
     */
    ToastMessageComponent.prototype.durationTimer;
    /**
     * @type {?}
     * @private
     */
    ToastMessageComponent.prototype.toastsService;
}
