// angular-bootstrap-toasts/esm5/lib/Components/toasts-container/toasts-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { AngularBootstrapToastsService } from '../../angular-bootstrap-toasts.service';
import { trigger, style, animate, transition } from '@angular/animations';
var ToastsContainerComponent = /** @class */ (function () {
    function ToastsContainerComponent(toastsService) {
        this.toastsService = toastsService;
        this.Placement = {
            position: 'topRight'
        };
        /**
         * Width as CSS value
         */
        this.Width = '250px';
        this.Classes = '';
        this.defaultMargin = '15px';
        this.MarginAsString = '';
        this.toastsList = [];
    }
    /**
     * @return {?}
     */
    ToastsContainerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.messagesListSubscription = this.toastsService.ToastsList$.subscribe((/**
         * @param {?} toasts
         * @return {?}
         */
        function (toasts) {
            _this.toastsList = toasts;
        }));
        this.initParams();
    };
    /**
     * @return {?}
     */
    ToastsContainerComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.messagesListSubscription) {
            this.messagesListSubscription.unsubscribe();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ToastsContainerComponent.prototype.initParams = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.Placement) {
            this.Placement = {};
        }
        this.Placement.marginTop = this.Placement.marginTop || this.defaultMargin;
        this.Placement.marginBottom = this.Placement.marginBottom || this.defaultMargin;
        this.Placement.marginLeft = this.Placement.marginLeft || this.defaultMargin;
        this.Placement.marginRight = this.Placement.marginRight || this.defaultMargin;
        this.Placement.position = this.Placement.position || 'topRight';
    };
    ToastsContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Angular-Bootstrap-Toasts-Container',
                    template: "<div\r\n    class=\"Advanced-Bootstrap-Container {{ Classes || '' }}\"\r\n    [style.width]=\"Width\"\r\n    [style.left]=\"(Placement.position === 'topLeft'      || Placement.position === 'bottomLeft')  ? Placement.marginLeft : ''\"\r\n    [style.right]=\"(Placement.position === 'topRight'    || Placement.position === 'bottomRight') ? Placement.marginRight : ''\"\r\n    [style.top]=\"(Placement.position === 'topLeft'       || Placement.position === 'topRight')    ? Placement.marginTop : ''\"\r\n    [style.bottom]=\"(Placement.position === 'bottomLeft' || Placement.position === 'bottomRight') ? Placement.marginBottom : ''\">\r\n    \r\n    <Toast-Message\r\n        *ngFor=\"let toast of toastsList\"\r\n        class=\"Toast-Message-Container\"\r\n        [@enterAnimation]\r\n        [Toast]=\"toast\"></Toast-Message>\r\n\r\n</div>",
                    animations: [
                        trigger('enterAnimation', [
                            transition(':enter', [
                                style({ transform: 'translateX(100%)', opacity: 0 }),
                                animate('350ms ease', style({ transform: 'translateX(0)', opacity: 1 }))
                            ]),
                            transition(':leave', [
                                style({ transform: 'translateX(0)', opacity: 1 }),
                                animate('300ms ease', style({ transform: 'scale(0.1)', opacity: 0 }))
                            ])
                        ])
                    ],
                    styles: [".Advanced-Bootstrap-Container{position:fixed;z-index:9999!important}"]
                }] }
    ];
    /** @nocollapse */
    ToastsContainerComponent.ctorParameters = function () { return [
        { type: AngularBootstrapToastsService }
    ]; };
    ToastsContainerComponent.propDecorators = {
        Placement: [{ type: Input }],
        Width: [{ type: Input }],
        Classes: [{ type: Input }]
    };
    return ToastsContainerComponent;
}());
export { ToastsContainerComponent };
if (false) {
    /** @type {?} */
    ToastsContainerComponent.prototype.Placement;
    /**
     * Width as CSS value
     * @type {?}
     */
    ToastsContainerComponent.prototype.Width;
    /** @type {?} */
    ToastsContainerComponent.prototype.Classes;
    /**
     * @type {?}
     * @private
     */
    ToastsContainerComponent.prototype.defaultMargin;
    /** @type {?} */
    ToastsContainerComponent.prototype.MarginAsString;
    /** @type {?} */
    ToastsContainerComponent.prototype.toastsList;
    /**
     * @type {?}
     * @private
     */
    ToastsContainerComponent.prototype.messagesListSubscription;
    /**
     * @type {?}
     * @private
     */
    ToastsContainerComponent.prototype.toastsService;
}
