// angular-bootstrap-toasts/esm5/lib/Models/toast-message.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
var ToastMessage = /** @class */ (function () {
    function ToastMessage(params, systemParams) {
        this.confirmationResult = new Subject();
        this.id = systemParams.id;
        this.type = systemParams.type;
        this.text = params.text;
        this.title = params.title;
        this.moment = params.moment;
        this.duration = params.duration;
        this.isProgressLineEnabled = params.showProgressLine;
        this.isDurationPausedByMouse = params.pauseDurationOnMouseEnter;
        this.isClosableByMouseClick = params.closeByClick;
        this.iconClass = params.iconClass;
        this.progressLineClass = params.progressLineClass;
        this.titleClass = params.titleClass;
        this.bodyClass = params.bodyClass;
        this.toastClass = params.toastClass;
        this.toolbarClass = params.toolbarClass;
        this.closeButtonClass = params.closeButtonClass;
        this.toolbarItems = params.toolbarItems;
    }
    Object.defineProperty(ToastMessage.prototype, "ConfirmationResult$", {
        get: /**
         * @return {?}
         */
        function () {
            return this.confirmationResult.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Id", {
        get: /**
         * @return {?}
         */
        function () {
            return this.id;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Type", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Title", {
        get: /**
         * @return {?}
         */
        function () {
            return this.title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Text", {
        get: /**
         * @return {?}
         */
        function () {
            return this.text;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Moment", {
        get: /**
         * @return {?}
         */
        function () {
            return this.moment;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "Duration", {
        get: /**
         * @return {?}
         */
        function () {
            return this.duration;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "ToolbarItems", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toolbarItems;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "IsProgressLineEnabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isProgressLineEnabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "IsDurationPausedByMouse", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isDurationPausedByMouse;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "IsClosableByMouseClick", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isClosableByMouseClick;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "IconClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.iconClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "CloseButtonClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.closeButtonClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "ProgressLineClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.progressLineClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "TitleClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.titleClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "BodyClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.bodyClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "ToolbarClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toolbarClass;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToastMessage.prototype, "ToastClass", {
        get: /**
         * @return {?}
         */
        function () {
            return this.toastClass;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} confirmationStatus
     * @return {?}
     */
    ToastMessage.prototype.Close = /**
     * @param {?=} confirmationStatus
     * @return {?}
     */
    function (confirmationStatus) {
        this.confirmationResult.next(confirmationStatus || false);
    };
    return ToastMessage;
}());
export { ToastMessage };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.confirmationResult;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.id;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.type;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.title;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.text;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.moment;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.duration;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.toolbarItems;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.isProgressLineEnabled;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.isDurationPausedByMouse;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.isClosableByMouseClick;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.iconClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.closeButtonClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.progressLineClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.titleClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.bodyClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.toolbarClass;
    /**
     * @type {?}
     * @private
     */
    ToastMessage.prototype.toastClass;
}
var ToastMessageParams = /** @class */ (function () {
    function ToastMessageParams() {
    }
    return ToastMessageParams;
}());
export { ToastMessageParams };
if (false) {
    /** @type {?} */
    ToastMessageParams.prototype.text;
    /** @type {?} */
    ToastMessageParams.prototype.title;
    /**
     * Great property to show time/date or something like this (short info near title)
     * @type {?}
     */
    ToastMessageParams.prototype.moment;
    /**
     * Time in milliseconds over which will be displayed toast
     * @type {?}
     */
    ToastMessageParams.prototype.duration;
    /**
     * Is toast will have progress line of duration time?
     * @type {?}
     */
    ToastMessageParams.prototype.showProgressLine;
    /**
     * Is toast will closable by mouse click on toast block
     * @type {?}
     */
    ToastMessageParams.prototype.closeByClick;
    /**
     * Is toast duration needs to be paused when mouse enter on toast block?
     * @type {?}
     */
    ToastMessageParams.prototype.pauseDurationOnMouseEnter;
    /**
     * Class of progress line
     * @type {?}
     */
    ToastMessageParams.prototype.progressLineClass;
    /**
     * Class of icon for title
     * For example: `fas fa-heart` (FontAwesome)
     * You can also add color class, ex: `fas fa-heart text-success` (Bootstrap color class)
     * @type {?}
     */
    ToastMessageParams.prototype.iconClass;
    /**
     * Class for title block of toast
     * @type {?}
     */
    ToastMessageParams.prototype.titleClass;
    /**
     * Class for body block of toast
     * @type {?}
     */
    ToastMessageParams.prototype.bodyClass;
    /**
     * Class for block of toast
     * @type {?}
     */
    ToastMessageParams.prototype.toastClass;
    /**
     * Class for toolbar block
     * @type {?}
     */
    ToastMessageParams.prototype.toolbarClass;
    /**
     * Class for close button
     * @type {?}
     */
    ToastMessageParams.prototype.closeButtonClass;
    /**
     * Buttons for toast with type "confirm"
     * @type {?}
     */
    ToastMessageParams.prototype.toolbarItems;
}
/**
 * @record
 */
export function SystemToastParams() { }
if (false) {
    /** @type {?} */
    SystemToastParams.prototype.id;
    /** @type {?} */
    SystemToastParams.prototype.type;
}
/**
 * @record
 */
export function ToastToolbarItems() { }
if (false) {
    /** @type {?} */
    ToastToolbarItems.prototype.actionButton;
    /** @type {?} */
    ToastToolbarItems.prototype.cancelButton;
}
/**
 * @record
 */
export function ToastToolbarButton() { }
if (false) {
    /** @type {?} */
    ToastToolbarButton.prototype.text;
    /** @type {?|undefined} */
    ToastToolbarButton.prototype.visible;
    /** @type {?|undefined} */
    ToastToolbarButton.prototype.class;
}
